### PR TITLE
Integrate Wheatley into Ringing Room as a 'bot'

### PR DESCRIPTION
**Pull request by kneasle**
This PR will, once complete, add [Wheatley](https://github.com/Kneasle/wheatley) to Ringing Room so that it can ring with people without them having to run it clunkily from a command line.  Wheatley is also currently not quite stable enough for use by potentially everyone, but hopefully by the time this PR is ready it will be.

Feedback and ideas for how to do this is very welcome - comment them on this PR and I'll see them.

### CODE REVIEW

**Comment by Kneasle** [comment]
Here's my idea for a strategy for how to do this.

- When a ringer clicks on the button to say they want a Wheatley to ring with them, the Flask server process spawns a new child process of Wheatley that runs independently, and gets passed `localhost` as a URL so that it talks directly to the currently running server process over WebSocket.

  Note that this doesn't spawn a child _thread_ - it spawns an entirely new _process_ with its own Python interpreter to minimise the possibility of Wheatley causing the main RR process to lag.

- Any time a user changes a Wheatley control, that is sent to the server as a WebSocket signal, which then gets broadcast to everyone in that room (including the Wheatley instance that is pretending to be a client).  The Ringing Room pages then update _their_ value of that parameter, and the Wheatley instance will also update its parameter value, and use that value from then on.

- When a user decides that they don't want Wheatley, that instance is killed (which is fine because it's completely independent from the RR instance).

I think this satisfies both requirements of minimising impact of Wheatley on the main RR experience when it isn't running, and requiring as little edits to both codebases as possible.

(By the way, I'm guessing that @lelandpaul, @BrynMarieR are OK with this change on the basis that you haven't closed it yet.  But if you are against it, then please say so :).  Feedback and ideas are also welcome.  Cheers!

**Comment by BrynMarieR** [comment]
Thanks -- we see this and your other PRs and we'll look through them tomorrow. Cheers!

**Comment by Kneasle** [comment]
Awesome - thanks!

I hope I haven't come across as impatient (that really wasn't the intention) - I just didn't want you to feel obliged to merge PRs just because I took the time to make them. Even waiting a week is way, way faster than PRs on big open source projects so I'm very cool with waiting a few days for you to review them.

**Comment by lelandpaul** [comment]
@Kneasle Thanks! We've talked about this, and are in full support — I think your architecture sounds good, and there's no doubt that this will add extremely value functionality to Ringing Room. Let us know when you're ready, and we'll a) make sure the UX is in line with our goals for the platform (not that I'm particularly worried!) and b) send it out to our team of beta testers before launching it.

**Comment by Kneasle** [comment]
**Cool**!  One thing that would be useful to know is: what OS does the Ringing Room servers run?  It shouldn't matter - Python _aught_ to be OS agnostic - but it would be useful to know (particularly if it's Windows) because I'll be spawning and killing child processes and there might be some OS specific stuff there.

Also, I envision potential issues with two people trying to set the same Wheatley setting at the same time, and I see two obvious solutions:
- The safest solution is to have one ringer in control of Wheatley so that there just _is_ no concurrency, but that's far from ideal since a) nothing in Ringing Room works this way and b) when they leave, all the Wheatley settings will be wiped or a janky succession system has to be made.  Add the fact that it's almost impossible to _actually_ detect when a WebSocket client leaves and this becomes extremely ugly to implement and use.
- The better solution, I think, is to have the continuous settings only update when the user releases them, and rely on the fact that if two people _do_ grab a control simultaneously, they'll probably both want to set it to nearly the same value and so the funky concurrency issues shouldn't actually occur enough to negatively damage the experience.

**Comment by BrynMarieR** [comment]
Sure thing --

- All of the servers are Linux (Ubuntu 18.04) -- so should be pretty standard
- I wonder if you can use something like what "host mode" uses. In "host mode" only the host(s) can access various controls -- this could include Wheatley. When the host(s) leave(s), everyone is given control of everything again. (but you're right, it's almost impossible to _actually_ detect when a WS client leaves, so this is risky. In practice we haven't heard of anyone having any issues.)
- I think it's fine to implement an egalitarian Wheatley (as it were) and we can do a beta-test session to see how well it works and test to see under what conditions we run into issues. It should be reasonably straightforward to change later if needed.

Are there any other security considerations? Is the only way to communicate with Wheatley the hotkeys, and the only way that Wheatley communicates back is generating changes from a pre-specified database of methods? (I'm wary of attackers using this to "Wheatley bomb" other towers... already technically a possibility)

**Comment by Kneasle** [comment]
Cool about using Linux - I do all my development on Ubuntu so I'll be able to test it reasonably well.

I think I'll implement the 'egalitarian' strategy for its simplicity, and change it if we run into issues later.

Wheatley instances will only be spawned by the `Tower` class with the URL and tower ID as an argument, so that won't be user accessible.  The only things that will be accessible to the user are what composition/method is being rung and some settings to do with sensitivity and ringing speed, the updates for which will be handled in the same way as ringing bells currently works.  So I don't think it adds any security flaws.

**Comment by BrynMarieR** [comment]
Seems great! Let us know when you'd like us to take a look.

**Comment by BrynMarieR** [comment]
An idea I've had to keep the simple interface of towers while still allowing Wheatley to show up and be useful -- forgive me if this was already the plan:

Can we "disable/enable" Wheatley in the tower controls for each tower, rather than adding directly to the tower page for all towers? Then Wheatley will show up as a user in the user list for towers in which they've been "enabled".

**Comment by kneasle** [comment]
My plan was to have Wheatley be 'active' by default (i.e. be present as a user, but not running on the server until he's actually needed), but able to be disabled if desired.  I figured that more people would try Wheatley if all you had to do was assign some bells and input a method name - but I'm slightly worried that it might come across as a bit pushy.

It would probably be a good plan to talk in person with you and Leland about this before we commit too much to a certain design - between time zones and my holiday sleep schedule there's usually a pretty long gap between replies on PR comments so an in-person discussion would be really useful.

**Comment by BrynMarieR** [comment]
We're moving to a new system for releases on git, using 'trunk-based' development. All new features will go on to the branch called `trunk' (once was `master,' but we're doing a re-name). Every release, we will branch off the currently-released version. When a new release comes, we'll branch off of trunk with some release-xx.xx and that code will be live on the site.

Feature branches work as usual; branch off of trunk, work, then merge back. Hopefully re-basing will not need to be too frequent.

**Comment by kneasle** [comment]
Noiiice!

I'm getting there with the Wheatley integration - most of my time has been swallowed by the necessary architectural changes to Wheatley rather than on changes to Ringing Room.  The last major hurdle is to decide how we're going to deal with someone changing the settings of the row generator whilst Wheatley is running, something it was never designed for.  But once we've decided on a solution, it shouldn't take too long to implement it.

**Comment by kneasle** [comment]
OK so progress update - we've passed a fairly major milestone in the process :smiley: - all the changes to Wheatley have been made, and we're now just waiting for @centreboard to review [the PR](https://github.com/kneasle/wheatley/pull/76) before it has its history cleaned and then is merged into Wheatley's master branch.  Now, the only outstanding changes are those related to making the user-facing side of Wheatley look pretty and user-friendly.  I've made a task checklist list in [this issue](https://github.com/kneasle/wheatley/issues/86) - if you have any more things that you want to add, comment it there or on this PR and I'll add it to the checklist.

This development work has brought up a few queries, though:

I've been working on adding an `Enable Wheatley` switch to the tower settings page (which is going OK), and I hit a bit of a blank about how the database works.  It appears that you can't just add columns to a SQLAlchemy database without it crashing when attempting to load the table that should have the column you added.  For my development server, this is fine because I can just wipe the database and start again, but how do you deal with this on your main servers - is there a 'right' way to add columns to the database?  How did you deal with adding host mode in this respect?

Also, what do you guys think is the best way of storing Wheatley's miscellaneous settings into the database?  My gut instinct is to just serialise them into JSON (as they would be sent over SocketIO) and store that in its own column of the `tower` table as a string - this way we don't have to migrate the database every time we change how Wheatley's settings are stored.

Thirdly, I need to implement a way of garbage collecting Wheatley instances because they are fairly resource hungry and if we let them pile up then it could badly impact the performance of the RR servers (which is critical for smooth ringing).  Wheatley instances are already killed when Python gcs either the `WheatleyInstance` object or the `Tower` object that owns it.  My idea for this is to garbage collect a Wheatley instance if no-one has rung in the tower it occupies for a certain period of time, e.g. half an hour or an hour (I don't think that waiting for everyone to leave the tower is very useful because of how easy it is for SocketIO clients to forget to leave towers).  I also thought that we might as well just GC the `Tower` objects like that (it seems that they currently get GCed after existing for 12 hours - is this right?), and then the Wheatley instance will get killed along with it.

Lastly, how (when implementing host mode) did you deal with people potentially turning host mode on or off whilst a `Tower` object is in existence? - I can't find a clear explanation just by surfing the code.  At any rate, I would like to be able to introduce Wheatley to a tower whist there are ringers in the tower so to that end we should decide how to handle adding or removing Wheatley from a `Tower` whilst it is in existence (because dis-/enabling Wheatley is not so simple as setting a flag in the `Tower` object - the Wheatley panel needs to be shown or hidden, and `Wheatley` needs to be added or removed from the users list).

**Comment by lelandpaul** [comment]
This is very exciting! Congrats on the milestone.

### Regarding database upgrades:

All database migration is handled through `flask-migration` (which is a wrapper on Alembic). The process is as follows:

1. Add the column to the SQLAlchemy object in `models.py`.
2. From the command line, run `> flask db migrate -m "<commit-name>"`, with some short-but-descriptive commit name. This will create a migration script in `migrations/versions/`; for a simple change like adding a boolean column, the default output should be fine.
3. From the command line, run `> flask db upgrade` to upgrade the database to that newest script.
4. Assuming everything works, commit the migration script.

### Regarding storing Wheatley settings

I like the idea of storing them as JSON in a single column. Let's go with that!

### Regarding garbage collection

I'll freely admit that the way Tower GC is currently handled is a bit hacky, and I'm open to suggestions. You're right that conditioning on when a bell was last rung seems like the best approach. I'll think about how to do this efficiently; definitely open to suggestions!

### Regarding enabling / disabling Host Mode / Wheatley

Changing the Host Mode setting currently works as follows:

``` python
# listeners.py
@socketio.on('c_host_mode')
def on_host_mode(json):
    # ...
    tower_id = json['tower_id']
    tower = towers[tower_id] # This is the *in memory* tower — creating it if need be
    tower.host_mode = json['new_mode'] # this setter automatically modifies the database
    # ...
```

However, you're right that at the moment we make no provision for what happens if Host Mode is enabled/disabled while people are in the tower — that is, it's currently possible to have a tower with multiple hosts where only some of those hosts would see the Host Mode toggle button. I've opened an issue to address this; thanks for bringing it to my attention!

I totally agree that it should be possible to enable Wheatley while there are ringers present in the tower. Two possible ways to handle this:

1. When Wheatley-status is changed, send a `s_wheatley_panel_enabled` (or similar); clients respond by toggling a show/hide variable on the Wheatley panel & adding or removing Wheatley from the userlist.

2. When Wheatley-status is changed, send a `s_redirect` to all clients pointing back to the same tower; this will force-refresh everyone currently in the tower.

I lean towards the former option — the latter has the danger that someone could force a refresh while a QP is being run in the tower. I'll let other people weigh in, though.

**Comment by kneasle** [comment]
> This is very exciting! Congrats on the milestone.

Thanks :grin:

---

> ### Regarding database upgrades:

Sweet! This seems to work just fine (though I seem to have made both columns nullable by accident, but I don't think there's any harm in always validating the stuff that comes out of a database so I'll change it if you think it's a big deal, otherwise I'll leave it).

---

> ### Regarding storing Wheatley settings
> 
> I like the idea of storing them as JSON in a single column. Let's go with that!

OK cool - that's what I'm going with now...

> ### Regarding garbage collection
> 
> I'll freely admit that the way Tower GC is currently handled is a bit hacky, and I'm open to suggestions. You're right that conditioning on when a bell was last rung seems like the best approach. I'll think about how to do this efficiently; definitely open to suggestions!

My gut reaction for implementing this would be to have the towers always store the last time a bell was rung (since setting local variables is really cheap, and doesn't involve timeouts or anything like that), and then every hour or so the server's main threads goes through all the towers and checks that stored time - and if enough time has passed then that tower is GCed.  I think this is both relatively simple and also doesn't severely impact the performance of the server all the time.  But this is far from the only strategy, and I'll leave it up to you to decide on what's best. 

---

> ### Regarding enabling / disabling Host Mode / Wheatley

Hmmm OK I think there might be some misunderstanding here.  There are two different things at play here (regarding Host Mode), and they both have really similar names.  So for sanity's sake, I'm just going to define what words I'll use:

1. There is a toggle in the tower settings that stores in the database whether or not the `Host Mode` button is allowed to appear in that tower.  I believe that this is called `host_mode_enabled` in the code, since this is the setter for that field (in `models.py`):
   ```python
   @host_mode_enabled.setter
   def host_mode_enabled(self, new_state):
       self._host_mode_enabled = new_state
       self.to_TowerDB().host_mode_enabled = new_state
       db.session.commit()
   ```
   I'm going to call this "host mode is **enabled** or **disabled**".

2. There is a setting in the `Tower` class that determines whether or not `Host Mode` is currently *active* in the tower.  This can only be set over SocketIO, and is never stored in the database.  I believe this is called `host_mode` in the code, since this is the setter for that field, which never touches the database:
   ```python
   @host_mode.setter
   def host_mode(self, new_mode):
       self._host_mode = new_mode
   ```
   I'm going to call this "host mode is _**active**_ or _**inactive**_".  Out of the 4 possible combinations of these two properties, the only disallowed state is being **disabled** but _**active**_. 

So, now we have got that out of the way, on to my plan for Wheatley (which is open for changes, but it's been working well so far).

## Wheatley

Firstly, the only aspect of Wheatley that the user has control over is whether or not Wheatley is **enabled**.  This will be a switch in the tower settings and will apply globally to the tower, causing the Wheatley panel to appear/disappear.  Wheatley will be then added as a user to that tower (regardless of whether or not a Wheatley instance is _**active**_), and a Wheatley instance will be _**activated**_ only when a user assigns him to a bell.  This way, we can avoid having Wheatley instances _**active**_ until the very last possible moment (thus avoiding the performance hit of _always_ having Wheatley _**active**_ for every tower with Wheatley **enabled**) whilst having it appear from the user that Wheatley is always there.  The accumulating Wheatley instances will then be GCed every so often, and their assignments reset (so that in order to make Wheatley ring, a user has to assign Wheatley to a bell and trigger the Wheatley _**activation**_ mechanic).

So, the question now becomes: suppose we have the following situation.  We have some ringers in a tower with Wheatley **disabled**.  Now, the owner of the tower goes to the tower's settings page, **enables** Wheatley, and returns to the tower.  They will see Wheatley being **enabled**, but everyone else will still see Wheatley being **disabled**.

The tricky part here is that the signal from the tower settings page to **activate** Wheatley is an HTTP request (I think...), which causes a switch to be set in the tower and the database to be saved.  And at this point, we have to send out a SocketIO signal to all the current users in the tower to make sure that they see the Wheatley panel appear.  But currently in the code, SocketIO signals are only sent in response to SocketIO - the HTTP listeners and the SocketIO listeners are in separate files, and so can't easily interact.  The only way I can think of to do this is to somehow hook a callback onto the HTTP code that calls a function inside `listeners.py` so that it has access to the `emit` function.  I don't really get how this part of the codebase works - how would you do something like this?

## Other questions
- How is the state of the tower (tower/hand sounds, bell number, etc.) initially broadcast to a new user?  Because the state of the Wheatley panel also needs to be broadcast, and I can't find a nice way to do it.
- What do you guys think is the best way to store Wheatley's settings in the database?  Because if we save the database every time a user changes a Wheatley setting on _any_ tower, then surely we'll overload the database pretty quickly (unless the database can deal with this?).

I think this is enough for now - comment size inflation seems to be a real thing and it's getting slightly out of hand :laughing:.

**Comment by lelandpaul** [comment]
I think we actually are on the same page about *activated* vs. *enabled* — in the codebase, the distinction for Host Mode is currently called *enabled* (= your activated) vs. *permitted* (= your enabled).

It's absolutely possible to have an HTTP callback emit to socketio. If you add `from app.extensions import socketio` to `routes.py`, then the socketio object will be available there, and `socketio.emit` will be able to send events to a socketio room; rooms are always named by the tower ID. So you should be able to send a `s_wheatley_enabled` or `s_redirect` (depending on which option you pursue) from an HTTP callback, causing users currently in a tower to see the Wheatley panel.

> - How is the state of the tower (tower/hand sounds, bell number, etc.) initially broadcast to a new user? Because the state of the Wheatley panel also needs to be broadcast, and I can't find a nice way to do it.

The current state is all broadcast in response to `c_join`. The listener for that event sends a number of events (like `s_audio_change`) to just the user who triggered the event. This is where Wheatley settings should get sent.


> - What do you guys think is the best way to store Wheatley's settings in the database? Because if we save the database every time a user changes a Wheatley setting on any tower, then surely we'll overload the database pretty quickly (unless the database can deal with this?).

Hm. So when we talked about this last, we discussed there being two kinds of settings: Anything that's likely to get changed multiple times per practice, and everything else. I don't see a problem with the latter sort (which should all be accessed via the Tower Settings page) hitting the database — I don't think it's likely to be often enough to be a problem. Everything that gets changed multiple times per practice, however, should not be written to the database at all. This is how e.g. audio type or tower size gets handled: It only goes to the `Tower` in-memory object, not the `TowerDB` database one, and thus these settings get lost when the tower is GCed. I think any Wheatley settings that wouldn't need to persist across GC shouldn't touch the database at all.

**Comment by kneasle** [comment]
Nice - I've got the Wheatley enabledness (by your definition `:)`) now completely working, complete with SocketIO broadcasts. I've gone the approach of having a `s_set_wheatley_enabledness` signal (for wont of a better name).  Broadcasting initial settings is also working.  Currently, I'm only storing Whealtey's enabledness in the database, and all other settings are stored on the in-memory `Tower` class and never hits the database.

However, I'm still confused by the use of `permitted`/`enabled`.  The tower settings page appears to call the process of allowing the `Host Mode` button to appear 'permitting host mode':
image
Whereas the code that is called when this is changed is called 'set_host_mode_enabled':
```python
@host_mode_enabled.setter
def host_mode_enabled(self, new_state):
    self._host_mode_enabled = new_state
    self.to_TowerDB().host_mode_enabled = new_state
    db.session.commit()
```
So I'm not sure which one to believe.  Unless I've got the wrong end of the stick, the code is also the opposite way round to what you said it was.

I'm going to keep using 'enabled' to refer to making the Wheatley panel appear or disappear since that makes this PR self-consistent.  Once we're happy with what it's going to be called, I can do a find/replace in only the code I've changed so that it's all correct.

**Comment by lelandpaul** [comment]
Whoops! Ok, I'm going to go through and standardize the code to match what you're doing — I think that's the best way to handle it. Thanks for bringing this to my attention.

**Comment by kneasle** [comment]
Cool! I'm not really bothered what we call everything, so long as the whole codebase is consistent.

I think this PR is nearing completion.  I've just done some heavy refactoring to the server-side code to collect all the Wheatley-implementation-detail-specific code into one file and class `app/wheatley.py` which will make future work and maintenance far easier.  There are still more things that need to be implemented but I think that this should be ready for beta-testing within the next week or so.

## The Wheatley panel

The Wheatley panel is, I think, basically done (the Debug settings are going to disappear - they are just for testing):
image

The method box has a snazzy autocomplete (kindly provided by Bob Wallis' [blueline website](https://rsw.me.uk/blueline/methods/)) which both allows you to only type a few letters of a method name, and also guarantees that the user cannot input an invalid method name:
image

If you change the stage then the server will try to move the currently-set method to that stage, falling back on even-bell Plain Bob if that is not defined (e.g. you went from `Cambridge Surprise Royal` to `Cambridge Surprise Minimus` which is not defined).

## Regarding merging of this PR
I can't help but notice that this PR is clocking up a) a huge number of commits and b) a large number of merge conflicts.

I think the best course of action is leave everything as-is until all the changes arrive as a result of the beta test.  Then, I'll do some aggressive history rewriting and boil down this crazy commit history into a handful of much larger commits, with no u-turns or fixes.  Then I'll rebase the branch to the latest `trunk`, dealing with the merge conflicts as they come.  After that, then the PR will contain only a proper commit history and be ready for merging.

How does that sound as a plan?

**Comment by kneasle** [comment]
Exciting news!

This PR is ready for beta testing :partying_face::tada::tada::tada:!

**Comment by BrynMarieR** [comment]
Great, we'll merge it in soon so it can go out on next Thursday's beta!

**Comment by kneasle** [comment]
Right.  Commit history has been aggressively cleaned, and all merge conflicts fixed.

So, here's how to get the correct version of Wheatley running on the ringing room server so that you can beta test it before we merge the Wheatley changes into `master`.

1.  Find somewhere to install the source code for Wheatley
2. `cd` to that directory (let's call it's path `$WHEATLEY_PATH` for this example)
3. Run `git clone --branch integrate-with-ringing-room https://github.com/kneasle/wheatley.git` to clone the correct Wheatley branch.
4. Set the environment variable `$WHEATLEY_CMD` to point to your local copy of [this file](https://github.com/kneasle/wheatley/blob/integrate-with-ringing-room/run-wheatley).  This should just be `$WHEATLEY_PATH/run-wheatley`.  `$WHEATLEY_CMD` tells the Wheatley integration what command to run when it wants Wheatley.
5. Do the database migration stuff
6. Boot up the server, go to a tower's settings page, and enable Wheatley (and enjoy!).

File: migrations/versions/18d1e622ccc5_add_wheatley_columns_to_database.py
Comment (centreboard) [inline migrations/versions/18d1e622ccc5_add_wheatley_columns_to_database.py:16]: I get `flask db upgrade`
> ERROR [root] Error: Multiple head revisions are present for given argument 'head'; please specify a specific target revision, '<branchname>@head' to narrow to a specific head, or 'heads' for all heads

 Does this need to be regenerated from the latest code?
```
flask db history
fd24fcbb6ba6 -> 18d1e622ccc5 (head), Add Wheatley columns to database
fd24fcbb6ba6 -> e1634351a997 (head), user tokens
22cc99c77bd3 -> fd24fcbb6ba6 (branchpoint), host mode enabled
ecf0969f42be -> 22cc99c77bd3, track tower created date, user joined date
```

File: app/templates/tower_settings.html
Comment (centreboard) [inline app/templates/tower_settings.html:118]: I think some text about to describe what Wheatley is would be useful. Host mode has a nice description along side it below.

File: app/static/ringing_room.js
Comment (centreboard) [inline app/static/ringing_room.js:1195]: I think it's the inline styles here that don't respect dark/light mode
image

File: migrations/versions/18d1e622ccc5_add_wheatley_columns_to_database.py
Comment (lelandpaul) [reply migrations/versions/18d1e622ccc5_add_wheatley_columns_to_database.py:16]: I recommend running `flask db merge heads`, which will effectively merge the two branches.

File: app/wheatley.py
Comment (centreboard) [inline app/wheatley.py:144]: With the change to user ids this should now be
```suggestion
        if call_name == "Look to" and USER_ID in self._tower.assignments.values():
```

File: app/static/ringing_room.js
Comment (centreboard) [inline app/static/ringing_room.js:1057]: I am blocked from fetching this in the background by the CORS policy on complib

image

File: app/static/ringing_room.js
Comment (centreboard) [inline app/static/ringing_room.js:965]: Consider having a minimum length (e.g. 3 characters) before searching for auto complete results

File: app/static/ringing_room.js
Comment (centreboard) [inline app/static/ringing_room.js:894]: Can we debounce these? Currently it sends a web request for every keystroke which is an unnecessary load on those hosting these apis.

**Comment by centreboard** [comment]
Would it be useful to have a feature flag for this so it can be merged and rolled out in a controlled manner? E.g. an environmental variable or some kind of global bool in a file that can be edited. 

There's already a good injection point in the Wheatley enabledness hiding the UI in the tower / stopping the process running, so I'm imagining a this setting will just force that to be false regardless of the database and hide/disable the tower settings UI.

File: migrations/versions/18d1e622ccc5_add_wheatley_columns_to_database.py
Comment (kneasle) [reply migrations/versions/18d1e622ccc5_add_wheatley_columns_to_database.py:16]: Yeah I also found that error and I think I got round it by specifying the exact commit I wanted to use.

File: app/static/ringing_room.js
Comment (kneasle) [reply app/static/ringing_room.js]: Fixed `:)`

File: app/static/ringing_room.js
Comment (kneasle) [reply app/static/ringing_room.js:1062]: Yeah I'm pestering Graham John to fix this - apparently he hasn't touched CORS for CompLib which seems plausible because he doesn't do much browser stuff outside CompLib itself.

File: app/wheatley.py
Comment (kneasle) [reply app/wheatley.py:144]: Nice spot - fixed!

File: app/templates/tower_settings.html
Comment (kneasle) [reply app/templates/tower_settings.html:118]: Yep it's on my to-do list

File: app/static/ringing_room.js
Comment (kneasle) [reply app/static/ringing_room.js:965]: I think it's fine - Bob Wallis' search refinement is excellent for giving you popular methods - for example, I know that I can type `c` then it enter and I'll get Cambridge on whatever stage.

File: app/static/ringing_room.js
Comment (kneasle) [reply app/static/ringing_room.js:894]: For Bob Wallis' blueline, he's agreed that this is totally OK - his website does this anyway so it's just like a few extra users on his site typing keystrokes.

For complib, I'll ask Graham John, but I think I'll probably end up debouncing this anyway.  I'd like to add functionality for allowing people to paste URLs that may include access keys to private compositions, but at the moment I just want to get Wheatley out there and basically functional before I need to start revising for exams.

**Comment by kneasle** [comment]
Here's a quick update of progress for this PR:

### Things done
- The Wheatley box now respects light/dark themes (thanks for @centreboard for pointing this out)
- I've fixed what I hope is all the fallout from rebasing the new changes to `trunk`, mainly the switch to assigning bells by user IDs.
- Thanks to Bob Wallis, the method search always produces plenty of suggestions, and 'non-standard' calls (for methods like Stedman) are now consistent with [blueline](https://rsw.me.uk/blueline/methods/).
- Wheatley now has a feature flag (Wheatley will only appear if the environment variable `$RR_ENABLE_WHEATLEY` is set to `1`).  This means that the changes don't need to be perfectly implemented before they can be merged. Also, if any game-breaking issues are found, setting `$RR_ENABLE_WHEATLEY` to anything other than `1` will remove Wheatley (even when the server is running).

### Things still to do
My 'master' checklist is [here](https://github.com/kneasle/wheatley/issues/86), but here are the main things that are still outstanding (but probably not necessary for merging or the beta test).

- [ ] Write a small explanation of what Wheatley is to go next in the tower settings
- [ ] Pester Graham John to enable CORS for RR (he's currently on holiday so not much can be done on that front).
- [ ] Put Wheatley in the most bullet-proof configuration possible for launch (even if this means removing fancy features)

Also, would it be useful for me to be around for the beta test? If it's not super late in England when you do it, I'm happy to come along to help with any questions/issues in real time.

File: app/static/ringing_room.js
Comment (kneasle) [reply app/static/ringing_room.js:1060]: Fixed

File: app/templates/tower_settings.html
Comment (kneasle) [reply app/templates/tower_settings.html:119]: Done.

**Comment by kneasle** [comment]
Right this is finally complete and ready to be merged.

**Review by BrynMarieR — APPROVED** [review]
We may make a few changes at some point (there's a typo or two in the comments; "enabledness" can be made more succinct by using conventional "enabled") but the overall structure looks really good. Thanks for all your hard work on this! beta coming soon!

**Pull request merged**
